### DISCUSSION TITLE
Improvements on ApplicationForm

### DISCUSSION THREAD

**Discussion by JamesTheClarke** [CntoDev/level-up#133] [Q&A]
Currently only the fields `Nickname` and `Email` are mandatory to fill in, all other fields can be skipped because they are either not enforced or pre-selected (case for DoB, TZ and pronouns).

The following fields should be mandatory to fill out for a complete application:

- [ ] `Where did you find us?` currently missing application field, but very important to know where our recruitment efforts should be focused on. Could be improved over our current enjin application form by giving six drop-down options: 
  - Reddit
  - Steam Forums
  - YouTube
  - BI Forums
  - Twitter
  - Other (user can give custom written response)
- [ ] `Language skill level` should be renamed to `What is your English skill level?` as most of us are non native English speakers - Language skill level is a bit misleading as it could be interpreted with regards to their native language
- [ ] `How much experience do you have playing with Arma series?` should be renamed to `How much experience do you have playing the Arma series? What was your first Arma game?` because "playing with Arma series" is grammatically a bit wonky.
- [ ] `Which Arma mods have you been using so far?`
- [ ] `Have you previously been a part of similar communities?` currently missing application field, this is often used by interviewers to get a sense of two things a.) what type of communities they have frequented before (milsim, role-play, casual etc.) and b.) if they are potentially a "community-hopper", some folks who list 5 or 6 units and are 17 or 18 are potentially getting kicked out repeatedly and move from place to place. Obviously a subjective measurement, but a good potential red flag indicator nonetheless to have for interviewers.
- [ ] `Where do you see yourself in the community, any specific roles or activities?`
- [ ] `Tell us a little about yourself.`
- [ ] `Owned DLCs (hold control for multiple selection)` should be expanded with a `none` option in order to make the field mandatory.

**Comment by enrico-ghidoni** [comment]
Merging `ApplicationForm` related feedback from issue list

**Comment by enrico-ghidoni** [comment]
From @JamesTheClarke 

Currently the Time Zone selection has a couple of issues which make it less userfriendly for applicants:

1.) Each single timezone has multiple representatives making the drop down field much longer than needed. E.g. UTC-7 is represented with 16 example cities. Each timezone should be represented with 1-3 examples max.

2.) Our community is mostly CET (UTC+1) / CEST (UTC+2) focused, yet the drag and drop field starts at UTC-11. Can we select by default CET / CEST and then have the drag and drop go up and down in sections?

3.) The Time zone selection menu should have a short explanation or question below it to clarify what we are asking. In our current application form on enjin this would be "What UTC time zone are you from?" as well as a link so people can find their timezone accurately: https://www.timeanddate.com/time/map/


Sidenote: does the Time Zone selection menu in the application form automatically update UTC according to summer time shifts? I.e. Central European Time (CET) is UTC+1 but Central European Summer Time (CEST) is UTC+2. Both are located in the Berlin/Paris region.

**Reply by enrico-ghidoni** [reply]
From @milivojm 

Hi!

[Timezone dropdown](https://github.com/CntoDev/rooster/blob/develop/src/Roster.Web/Areas/Roster/Pages/ApplicationForm/Apply.cshtml#L27) is created using data from [here](https://github.com/CntoDev/rooster/blob/develop/src/Roster.Web/Areas/Roster/Pages/ApplicationForm/Apply.cshtml.cs#L185). This in turn is filled from https://docs.microsoft.com/en-us/dotnet/api/system.timezoneinfo.getsystemtimezones?view=net-6.0

**Reply by enrico-ghidoni** [reply]
From @milivojm 

> Sidenote: does the Time Zone selection menu in the application form automatically update UTC according to summer time shifts? I.e. Central European Time (CET) is UTC+1 but Central European Summer Time (CEST) is UTC+2. Both are located in the Berlin/Paris region.

It's unlikely. See https://stackoverflow.com/questions/2961848/how-to-use-timezoneinfo-to-get-local-time-during-daylight-saving-time for more.

**Reply by enrico-ghidoni** [reply]
From @JamesTheClarke 

> [Timezone dropdown](https://github.com/CntoDev/rooster/blob/develop/src/Roster.Web/Areas/Roster/Pages/ApplicationForm/Apply.cshtml#L27) is created using data from [here](https://github.com/CntoDev/rooster/blob/develop/src/Roster.Web/Areas/Roster/Pages/ApplicationForm/Apply.cshtml.cs#L185). This in turn is filled from https://docs.microsoft.com/en-us/dotnet/api/system.timezoneinfo.getsystemtimezones?view=net-6.0

Is it possible for us to pull that data and then pre-select UTC+1 as the default option? Because as is the vast majority of our users will currently have to scroll down from UTC-11 to UTC-1, UTC+0, UTC+1 or UTC+2. Considering the amount of different options for each UTC in the list that's a lot of unecessary scrolling we could optimise for them.

**Reply by enrico-ghidoni** [reply]
From @milivojm 

Yes, it is. It's also possible to filter out unwanted time zones.

**Comment by enrico-ghidoni** [comment]
From @JamesTheClarke 

The nickname field in the application could be improved with the following changes:

- [ ] The explanation text should change from `Use the same nickname as in Enjin` to `This will be your uniform nickname for all CNTO platforms. Limited to max. 2 words in length, each capitalised, max. 10 characters in total (spaces included), avoid odd formatting.`
- [x] The field should automatically check for the correct guidelines and give a red error message if the chosen nickname does not comply with our naming convention.


Detailed nicknaming convention:

- length of max 10 characters (spaces included)
- can include max 1 space
- can not include any numbers
- can not include any symbols / special characters
- can not be formatted with odd / hard to read capitalisation (e.g. CoWbOyKiLlAh)
- can not be all caps if longer than three characters (e.g. CLARKE) 
- 1st letter of nickname needs to be capitalised, if the nickname is 2 words then both need to be capitalised

**Reply by enrico-ghidoni** [reply]
From @enrico-ghidoni 

The help text was changed to `Use the same nickname as in Enjin` for UI reasons. Here's how it would look with the full text and in my opinion it's just ugly. So since the naming conventions are already enforced on Enjin it would be useless to point them out again and I figured telling applicants to use what is an already validated nickname has the same effect without compromising the whole form look.

![image](https://user-images.githubusercontent.com/8669492/153175808-73835f20-7f99-45c5-96db-53e37295ca9a.png)

Concerning nickname validation, it already happens and all the checks are in place. It's not asynchronous (as in, the nickname is verified once you submit the form and not when you type in the nickname itself).
The only missing check is the odd capitalization as it's very subjective. Instead, we can force it to be always in lower case except for the first letter which must be capitalized (which is already enforced).

**Reply by JamesTheClarke** [reply]
I agree that the suggested text is too large. I that case I still propose to change it to `This will be your uniform nickname for all CNTO platforms.` because `Use the same nickname as in Enjin` references a.) a site we might leave at some point in the future and b.) the nickname has to be identical for all CNTO services, not just the Rooster and Enjin.

Then the automatic validation can handle the rest, tested a few combinations and the error message was always easy to understand, so it should suffice on it's own without further help text.

Regarding odd capitalisation: as long as Interviewing Manager and SSgts can manually adjust their nickname on the Rooster we'll be fine.

**Comment by enrico-ghidoni** [comment]
From @JamesTheClarke 

https://rooster.carpenoctem.co/Roster/ApplicationForm/ListApplications

Sidenote: I'd rename `List applications` to `List of Applications`.

Main issue: currently there is no tracking of who accepted or rejected applications and why. The following features would be great to keep similar records to our current enjin application process.

- [ ] it needs to be visible who accepted or rejected an application.
- [ ] it needs to be visible when an application was accepted or rejected
- [ ] when an application is rejected a reason should be given and tracked as a comment.
- [ ] Interviewers, Corporals and SSgts need to be able to leave private comments (not visible to the applicant) on each application. This is used to identify red flags collectively before the interview starts, thus the interviewer has all the necessary information in one single location. It also serves to leave comments in cases one doesn't hear from the applicant in time. Examples:
  - https://www.carpenoctem.co/dashboard/applications/application?app_id=10019820
  - https://www.carpenoctem.co/dashboard/applications/application?app_id=10003487
  - https://www.carpenoctem.co/dashboard/applications/application?app_id=10009590
- [ ] Interviewers should be able to leave public comments visible to the applicant. This is used to inform them either a.) that the application meets our standards and they should contact us on TS or Discord for and interview or b.) it does not meet our standards and has been rejected outright.

**Reply by enrico-ghidoni** [reply]
From @enrico-ghidoni 

Activity logs are a planned feature, see #131. The core idea is that every action performed either by users or by the system will get reported to allow for better tracing of *who did what* and *when*.

From there it will be trivial to add those details to the list of application forms.

Concerning private and public comments on the application form, we'll keep this on the to-do list. I'm wondering whether it's necessary for replacing the old Roster or not.

**Reply by enrico-ghidoni** [reply]
From @JamesTheClarke 

> Concerning private and public comments on the application form, we'll keep this on the to-do list. I'm wondering whether it's necessary for replacing the old Roster or not.

This feature would definitely still be very appreciated. The onboarding process isn't always handled by the same person at the three different stages. Thus it can be very helpful to look up what another person has observed during an Interview for example or what kind of redflags were noticed in the application. For record keeping it was often also useful in the past when an escalation occured and I was able to check if anyone noticed something early on or if that person kinda flew under the radar.

Currently enjin has some limitations in collaborative onboarding, thus interviewers mix the roster, the application comment section the interviewing discord channel and the forums to share information, however that means sometimes you have to look in various areas to get up to speed. If possible it would be ideal if the new roster could track the various stages of the onboarding process and become the single point of information storage / sharing for all involved staff members.

**Reply by enrico-ghidoni** [reply]
From @enrico-ghidoni 

Don't get me wrong I agree that streamlining the whole workflow is the end-goal of Rooster. I'm just trying to filter out features that are critical for v1 and those that will be a good addition later down the line.

I noticed old Roster has a "Notes" section for each recruit (and member in general). Not a big fan of generic note text. We can use that a temporary solution if required and move to a more structured system as you mentioned in your first message.

**Reply by JamesTheClarke** [reply]
I can agree with the private comments for a later version, though I still maintain that any new solution for our application process should ideally offer the same functionality as our current process before roll out - this would not be a new feature put on hold but an existing one temporarily removed which could impact acceptance from current users.

~~Regarding public comments: currently on enjin they are a must, it's how we inform the applicant how to proceed to the interview stage. We never accept an application form before they passed the interview - this is how we track without any additional lists that a person has cleared both the written application and the interview stage. However, with the new rooster we can adjust the process by sending out the required interview information to the applicant via Email by accepting the form if it looks alright. For interviewers however, it would be extremely useful to have several tickboxes to track the status of the applicant without the need of excel lists or as it is currently done with the status-thread on the forums~~

- `Interview completed`
- `Mod assessment completed`
- `Bootcamp completed`
- `Postive evaluation` (this manual tickbox would would update the status from **Recruit** to **Reservist** - should be restricted to Interview Manager or SSgts (backup) only to avoid accidental promotions, should only be clickeable if the applicant attended at least five operations)

The last restricted tickbox is a nice-to-have feature for the deluxe version of the Rooster. In the beginning in could just be a rudimentary tickbox tag to complete the recruitment process.

EDIT: scratch that, just accepted another test acccount and realised that those labels already exist.

EDIT2: is there an undo-button for misclicks? all the label buttons are very close to each other and it can be easy to click on the wrong button by accident.

EDIT3: there's no `Interview completed` label button. At what stage are interviewers supposed to accept the application form in the new rooster? after the interview or before (if the application "looks good")? If it's after the interview then how is the applicant informed that they are clear to proceed to the next stage? This is currently handled with a public comment on the enjin application, they get a public comment (with automated enjin notification) that their application looks good and they can proceed to contact an interviewer for the interview.

**Comment by enrico-ghidoni** [comment]
From @JamesTheClarke 

When selecting `Other` one cannot specify a custom value even though the explanation text states it is possible:

`By picking "Other" you can specify a custom set of pronouns.`

**Reply by enrico-ghidoni** [reply]
From @enrico-ghidoni 

This wasn't implemented due to lack of time and focus being on other features. Personally I'd take out the help text (the one saying that by picking "other" you can specify the pronouns) for the time being as it's not a critical feature.

What could fill the gap before the feature is implemented? Interviewer can ask the applicant what the preferred pronouns are and outline them in the recruit feedback on Enjin.

**Reply by JamesTheClarke** [reply]
I agree that this could be done manually via the Recruit feedback thread by the interviewer, however, the reason we implemented it in the form is because sometimes sharing vital information like that gets forgotten, either by the interviewer asking during the interview stage or them forgetting to post the information later in the relevant thread. Not to mention that sometimes people post stuff on Discord instead of the feedback thread and information gets spread thin across several channels.

In short, while doable via the forums it re-introduces several potential manual failstages that could be avoided by adding a custom text field. If possible we should include this for v1 to avoid headaches in the long run as it is an existing feature of our current enjin application form.

**Comment by enrico-ghidoni** [comment]
From @JamesTheClarke 

In my view the Application Form should be limited to what we need to initially process a recruit. If the form is too lenghty it can be intimidating and potentially deter people from filling it out.

The following field should not be present in the application but instead be able to be edited by interviewers, interviewing manager, member manager and ssgts in the member profile during different stages of the onboarding process (similar to how interviewers currently add the BI account username after mod assessment & bootcamp):

- `Bohemia Interactive nickname` should be renamed to `Bohemia Interactive Account Username` as this is what BI references on their Arma 3 Unit Page.
- `Discord username` should be renamed to `Discord Profile ID`, usernames can change, the ID is a unique identifyer.
- `GitHub nickname` should be renamed to `GitHub Account Name`, the nickname could be misleading because you can give yourself a nicer looking profile nickname. Example my profile's nickname is "Clarke", my account name is "JamesTheClarke": https://github.com/JamesTheClarke
- `Google e-mail`should be renamed to `Gmail Address`
- `Steam username` should be renamed to `Steam ID` for the same reason as with GitHub / Discord, one can change the username, but the ID (17-digit number) is permanent.
- `Teamspeak ID` should be renamed `TeamSpeak ID` just to keep things uniform (see GitHub).

This would cut six fields from the application form and the most important three (Discord, BI Account and TS ID) can be filled in by Interviewers at different stages of the onboarding process.

**Comment by JamesTheClarke** [comment]
Minor bug:

I've been testing out different combinations of Nicknames to ensure the automated check response are accurate so the end user knows what they have to change. The checks held true for every combination except one: managed to get this through with the wrong automated response.

1. Type Username `Test test` (this would not be allowed as per our naming convention two words have to be capitalised)
2. System will respond with `Application form validation failed. Nickname cannot contain numbers or symbols.` which is incorrect. It should say instead `Application form validation failed. Individual words in a nickname need to be capitalised.`

![screenshot-rooster carpenoctem co-2022 02 14-10_13_11](https://user-images.githubusercontent.com/11644347/153834501-bc99e4b7-7915-49d0-a297-17111f78e14e.png)

**Reply by enrico-ghidoni** [reply]
Possibly caused by the "letters only" condition in [`src/Roster.Core/Domain/MemberNickname.cs:29](https://github.com/CntoDev/rooster/blob/3e01dbc3f82b75847ef4bdac3ec9dcf249a28515/src/Roster.Core/Domain/MemberNickname.cs#L29).

Should be updated to allow for white spaces as the white space maximum condition precedes the one under consideration.

**Comment by enrico-ghidoni** [comment]
Gonna summarize the proposed changes here so we have an overall view of how big of a task this could be as well as the decision to either implement the change before `v1` or not.

### 1 [Missing form fields and rewording of existing ones](https://github.com/CntoDev/rooster/discussions/133#discussion-3865788)
 1. Addition of new mandatory form fields
 2. Rewording of existing form fields
 3. Addition of `none` option under the `Dlcs` field

### 2 [Making the `TimeZone` field more user friendly](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2141871)

### 3 [Change the wording on `Nickname` field help text](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2141873)

### 4 [Integrating user actions and logging with the ApplicationForm](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2141881)
 1. Log user actions for form acceptance/rejection
 2. Add a `reason` field upon rejecting an application
 3. Add private comments visible only to users
 4. Add public comments

### 5 [Finish implementing custom pronouns](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2141899)

### 6 [Remove fields from ApplicationForm, make them fillable by interviewer at a later stage](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2141902)

### 7 [Fix white space bug in `Nickname` field](https://github.com/CntoDev/rooster/discussions/133#discussioncomment-2171609)
